### PR TITLE
Ronan Kelly

### PR DESCRIPTION
**Pull request by RonanKel**

### CODE REVIEW

File: brevets/acp_times.py [hunk 1]
```diff
@@ -26,7 +31,48 @@ def open_time(control_dist_km, brevet_dist_km, brevet_start_time):
...
+        hours += 200 / 34
+    elif (brevet_dist_km == 300 and control_dist_km >= 300):
+        hours += 200 / 34
+        control_dist_km -= 200
+        hours += 100 / 32
+    elif (control_dist_km <= 400):
+        hours += 200/34
+        control_dist_km -= 200
+        hours += control_dist_km/32
+    elif (brevet_dist_km == 400 and control_dist_km >= 400):
```
Comment (Reverb-sudo) [inline brevets/acp_times.py:50]: This is great logic and gives the correct output. However, it is quite lengthy and and doesn't abide by "DRY", in that you can factor a lot of these out. You can make a list of tuples and how they relate to each other. For example, you can set [(200,200,34),(400,200,32)...] where we have (bound,length,maxspeed). You can base your logic on that. However, your code makes the logic quite explicit. Good work here.

File: brevets/flask_brevets.py [hunk 2]
```diff
@@ -44,21 +45,31 @@ def page_not_found(error):
...
     app.logger.debug("km={}".format(km))
     app.logger.debug("request.args: {}".format(request.args))
     # FIXME!
     # Right now, only the current time is passed as the start time
     # and control distance is fixed to 200
     # You should get these from the webpage!
-    open_time = acp_times.open_time(km, 200, arrow.now().isoformat).format('YYYY-MM-DDTHH:mm')
-    close_time = acp_times.close_time(km, 200, arrow.now().isoformat).format('YYYY-MM-DDTHH:mm')
+    open_time = acp_times.open_time(km, distance, begin_date_arrow).format('YYYY-MM-DDTHH:mm')
+    close_time = acp_times.close_time(km, distance, begin_date_arrow).format('YYYY-MM-DDTHH:mm')
```
Comment (Reverb-sudo) [inline brevets/flask_brevets.py:72]: I like how correct this change was.

File: brevets/acp_times.py [hunk 3]
```diff
@@ -41,4 +87,30 @@ def close_time(control_dist_km, brevet_dist_km, brevet_start_time):
        An arrow object indicating the control close time.
        This will be in the same time zone as the brevet start time.
     """
-    return arrow.now()
+    hours = 0
+    if (control_dist_km == 0):
+        hours = 1
+    elif (control_dist_km >= brevet_dist_km and control_dist_km <= brevet_dist_km * 1.2):
+        hours = time_limits[brevet_dist_km]
+    elif (control_dist_km < 60):
```
Comment (Reverb-sudo) [inline brevets/acp_times.py:95]: Great acknowledgement to the oddity here.

File: README.md [hunk 4]
```diff
@@ -1,66 +1,25 @@
...
-> That's *"controle"* with an *e*, because it's French, although "control" is also accepted. Controls are points where a rider must obtain proof of passage, and control[e] times are the minimum and maximum times by which the rider must arrive at the location.
+	This application lets us run a web program using flask. This program helps us to calculate a race's (brevet's) control's start and open time without having to reload the page.
+To start this program all you must do is enter the '/brevets' directory and type "python3 flask_brevets.py". This will read the Dockerfile start it and let you access the server through which ever port
+you have described in your credentials.ini file. In your browser type 'localhost:<port>'.
 
-### ACP controle times
+	The way this works is that you select a start date, time, brevet length, and at what distance you would like there to be a control. Using JS and flask every time you enter in a control position,
+the JS will get a getJSON request to the flask_brevets.py file. It will send the data of the start date and time, brevets distance, and control position. Then the script will pass it through
```
Comment (Reverb-sudo) [inline README.md:10]: Nice, natural, explanation here which complements the code superbly.

File: README.md [hunk 5]
```diff
@@ -1,66 +1,25 @@
...
+that will have passed since the start time, then truncates the number of minutes so it doesn't round and then returns an arrow object which now represents the time in which the control opens.
+It is tested on whether distance is 0, less than 200, 300, 400, 600, and 1000, then calculates. Or it tests to see if the control is past the brevet length in which it will return the same time the
+control at the control that is at the brevet length.
 
-We are essentially replacing the calculator here [https://rusa.org/octime_acp.html](https://rusa.org/octime_acp.html). We can also use that calculator to clarify requirements and develop test data. 
 
-## Tasks
+	For the close times function, it tests to see if the control is at 0, if its less than 60, 600, and 1000 then calculates the hours and returns an arrow object like the open times function. It also tests to see if a control is past the brevet length and then returns the same closing time a control at the brevet length would be. These function largely the same, however the closing time function uses
+dictionary based data to check how long a brevet max closing time would be.
```
Comment (Reverb-sudo) [inline README.md:20]: I think either function could use dictionary data. I assume you meant to have the implementation dictionary-based. Other than that, this is a great explanation!

File: brevets/tests/test_acp_times.py [hunk 6]
```diff
@@ -4,8 +4,81 @@
...
                     level=logging.WARNING)
 log = logging.getLogger(__name__)
+
+
+
+def assert_loop(dist, checkpoints, start_time):
+    for km, time_tuple in checkpoints.items():
+        checkpoint_open, checkpoint_close = time_tuple
+        assert open_time(km, dist, start_time) == checkpoint_open
+        assert close_time(km, dist, start_time) == checkpoint_close
```
Comment (Reverb-sudo) [inline brevets/tests/test_acp_times.py:26]: Great function to factor out the assertions, making the code more concise, I definitely should have done this!

File: brevets/tests/test_acp_times.py [hunk 7]
```diff
@@ -4,8 +4,81 @@
...
+    assert_loop(dist, checkpoints, start_time)
+
+def test_brevet2():
+    start_time = arrow.get("2022-02-22 00:00", "YYYY-MM-DD HH:mm")
+    dist = 400
+    checkpoints = {
+        20: (start_time.shift(hours=0,minutes=35), start_time.shift(hours=2,minutes=0)),
+        250: (start_time.shift(hours=7,minutes=27), start_time.shift(hours=16,minutes=40)),
+        400: (start_time.shift(hours=12, minutes=8), start_time.shift(hours=27, minutes=0)),
+        410: (start_time.shift(hours=12, minutes=8), start_time.shift(hours=27, minutes=0))
```
Comment (Reverb-sudo) [inline brevets/tests/test_acp_times.py:47]: Great edge case to test for clipping.

**Review by Reverb-sudo — COMMENTED** [review]
You have correct logic and it looks like the overall website and code works correctly. I see that you may have attempted the dictionary-based logic, which would have only made it a bit better. But everything else is well thought-out and purposeful. Well done!
